server.go

```diff
@@ -21,6 +46,70 @@ type Server struct {
 	PublicKeyHandler    PublicKeyHandler    // public key authentication handler
 	PtyCallback         PtyCallback         // callback for allowing PTY sessions, allows all if nil
 	PermissionsCallback PermissionsCallback // optional callback for setting up permissions
+
+	// Internal fields. Note that the zero value for these should be a state we
+	// can detect so the Server can still be instantiated using &Server{}.
+
+	listener net.Listener    // listener currently being used by Serve, nil otherwise
+	mu       sync.Mutex      // general lock around the Server state
+	wg       *sync.WaitGroup // WaitGroup to track the number of connections still running
+	state    serverState     // tracks if the server is running or not
+	doneChan chan struct{}   // tracks if we want to exit
+	killChan chan struct{}   // tracks if we're blindly killing connections
+}
+
+// getKillChan is used to get or create the kill channel
+func (srv *Server) getKillChan() chan struct{} {
```

Comment:
This is also similar to what `net/http` does. The locks and such are used more there because it's more complicated, but the general idea is the same. A large portion of this takes inspiration from there.